core/src/main/java/com/scalar/db/storage/jdbc/RdbEngineStrategy.java

```diff
@@ -131,4 +143,56 @@ default String getPattern(LikeExpression likeExpression) {
   default @Nullable String getSchemaName(String namespace) {
     return namespace;
   }
+
+  default Object encodeDate(DateColumn column) {
+    assert column.getDateValue() != null;
+    return column.getDateValue();
+  }
+
+  default Object encodeTime(TimeColumn column) {
```

Comment:
Using Object type is hard for me to confirm if the implementation returns correct types. How about using generics like this?

```java
public interface RdbEngineStrategy<T_DATE, T_TIME, T_TIMESTAMP, T_TIMESTAMP_TZ> {

  ...

  T_DATE encodeDate(DateColumn column);
  T_TIME encodeTime(TimeColumn column);
  T_TIMESTAMP encodeTimestamp(TimestampColumn column);
  T_TIMESTAMP_TZ encodeTimestampTZ(TimestampTZColumn column);
}

class RdbEngineSqlServer implements RdbEngineStrategy<String, LocalTime, String, DateTimeOffset> {
  ...

  @Override
  public LocalTime encodeTime(TimeColumn column) {
    return column.getTimeValue();
  }

  @Override
  public String encodeDate(DateColumn column) {
    return column.getDateValue().format(DateTimeFormatter.BASIC_ISO_DATE);
  }

  ...
}
```

With this we can detect wrong type usage at compile time.

Reply:
I like the idea of enforcing type but then the default implementation for the method in the interface cannot added.
So, this means we would need to have these methods in each `RdbEngine` even when the default behavior does not need to be overridden.  
```
  @Override
  public LocalDate encodeDate(DateColumn column) {
    assert column.getDateValue() != null;
    return column.getDateValue();
  }

  @Override
  public LocalTime encodeTime(TimeColumn column) {
    assert column.getTimeValue() != null;
    return column.getTimeValue();
  }

  @Override
  public LocalDateTime encodeTimestamp(TimestampColumn column) {
    assert column.getTimestampValue() != null;
    return column.getTimestampValue();
  }

  @Override
  public OffsetDateTime encodeTimestampTZ(TimestampTZColumn column) {
    assert column.getTimestampTZValue() != null;
    return column.getTimestampTZValue().atOffset(ZoneOffset.UTC);
  }
```
We gain type safety but we lose on the readability and increase code duplication. Do you think that the pros outweigh the cons ?

Reply:
As you said, it's a trade-off. So, this might be a matter of taste, but I would take the type safety. Also, I think the code readability with the generics version is better for me. The code duplication is actually a drawback, but the logic and implementation are simple and I don't think it's a significant issue.

But, yeah it's a trade-off choice. I don't insist on it, and it's up to you.

Reply:
Alright, I updated it in https://github.com/scalar-labs/scalardb/pull/2437/commits/be71af53c8fea29cee89a6a177b67cb763be6377